### PR TITLE
Rearrange hash key responsibilities

### PR DESCRIPTION
**Pull request by BrianHicks**
I was driving the other day and sort of mulling over the architecture of the system, and realized that having `Job` totally own its key based on external data is only going to get more annoying when we add recursive dependencies. A simpler architecture would be to let the coordinator coordinate the right key all in one spot.

So where before we would ask Job to calculate the hash for both its direct fields and pass in a Dict of input file hashes, now we ask Job to calculate a key for only the direct fields, and then build on it to get the full hash.

I also went ahead and documented the approach! Should pair well with #46, although these PRs don't depend on each other.

### CODE REVIEW

**Comment by BrianHicks** [comment]
@bhansconnect or @rtfeldman, any chance one of y'all could look at this sometime this week? It's ready to go (and I've just resolved the merge conflict.)

File: docs/internals/how-we-determine-when-to-run-jobs.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+First of all, mtime comparisons causes problems in unintuitive ways ([as documented by Avery Pennarun](https://apenwarr.ca/log/20181113)).
+To sum up: you can't safely assume that mtime will only increase, or increase in a way that's always reliable to know if you need to rebuild.
+
+You can get a pretty good idea by combining mtime with other information, though!
+For example, we definitely need to rebuild if a file's length changed.
+We look at more than just that, though: in addition to length and mtime (calculated on all systems), we also read the inode number, permissions, UID, and GID on Unix-family systems (e.g. Linux and macOS.)
+
+"Add more metadata" is good enough to avoid the problem of not rebuilding when a file changes, but we can still get unnecessary rebuilds when the file metadata changes without the content changing (e.g. `touch -m some-file`.)
+To avoid this (and be able to skip as many rebuilds as possible) we also hash all the files.
+It would be unacceptably slow to recalculate hashes for file on every run, though, so we cache them according to a key derived from the metadata.
```
Comment (bhansconnect) [inline docs/internals/how-we-determine-when-to-run-jobs.md:50]: I am not sure I understand, If you are hashing just the metadata, how does that tell you if the file actually changed or not? As you said, the metadata could have an mtime change that is not an actual content change.

File: src/coordinator.rs [hunk 2]
```diff
@@ -148,7 +162,7 @@ impl Builder {
         let file_hashes = File::create(file_hashes_path)
             .context("could not open file hash cache to store new hashes")?;
         // TODO: BufWriter?
-        serde_json::to_writer(file_hashes, &coordinator.meta_to_hash)
+        serde_json::to_writer(file_hashes, &meta_to_hash)
```
Comment (bhansconnect) [inline src/coordinator.rs:165]: Does it matter that this order could theoretically change? Like to meta_to_hash with the same items but different order.

File: src/job.rs [hunk 3]
```diff
@@ -2,32 +2,78 @@ use crate::glue;
...
+    pub fn based_on(id: &Key<Base>) -> Self {
+        let mut builder = Self::new();
+        id.hash(&mut builder.0);
+
+        builder
+    }
+
+    pub fn add_file(&mut self, path: &Path, content_hash: &str) {
+        path.hash(&mut self.0);
+        content_hash.hash(&mut self.0);
```
Comment (bhansconnect) [inline src/job.rs:30]: I thought we were avoiding hashing file content and just hashing the metadata, but I am probably just confused from comments elsewhere to use here.

**Review by bhansconnect — APPROVED** [review]
Looks good. Still a bit confused about some metadata comments and when you do or don't want to hash the file content, but otherwise this looks great.

File: docs/internals/how-we-determine-when-to-run-jobs.md (hunk 1 above)
Comment (BrianHicks) [reply docs/internals/how-we-determine-when-to-run-jobs.md:50]: we're not hashing the metadata, but the content of the file. It's just that we re-hash the file anytime the metadata key changes.

File: src/coordinator.rs (hunk 2 above)
Comment (BrianHicks) [reply src/coordinator.rs:165]: nah, this one's fine. It's just used as a KV store. Another one of those places that'll move to a better database eventually.

**Pull request merged**